File: 5.10/Capacity_Refinement.m
function B_Final = Capacity_Refinement(B_MWC, W_net, Tile_Size, dependency_matrix, C)
% Capacity_Refinement - 阶段三：处理容量超限时的依赖感知削减。
% 
% 输入:
%   B_MWC (1xn logical): MWC 算法得到的最优集合 (可能超限)。
%   W_net (1xn): 瓦片的净预期收益向量。
%   Tile_Size (1xn): 瓦片的大小 (这里假设是1)。
%   dependency_matrix (nxn): 依赖关系矩阵 (i -> j 表示 i 依赖 j)。
%   C (标量): RSU 的最大存储容量。
%
% 输出:
%   B_Final (1xn logical): 满足容量约束和依赖约束的最终缓存决策。

num_tiles = length(W_net);
B_Current = B_MWC; % 从 MWC 结果开始
CurrentSize = sum(B_Current .* Tile_Size);

if CurrentSize <= C
    B_Final = B_Current;
    return;
end

% 1. 重新计算效用密度 Psi[i, b]
% 效用密度 Psi[i, b] = 净预期收益 W(b) / 瓦片大小 Size[b]
% 瓦片大小 Tile_Size[b] 在这里就是 Tile_Size
Psi = W_net ./ Tile_Size;

% 2. 识别需要排序的瓦片 (仅限于当前缓存集合 B_Current)
Indices_To_Consider = find(B_Current);
Psi_to_sort = Psi(Indices_To_Consider);

% 3. 按效用密度降序排序
[~, Sorted_Indices_Local] = sort(Psi_to_sort, 'descend');
Sorted_Indices_Global = Indices_To_Consider(Sorted_Indices_Local);

% 4. 依赖感知削减 (贪婪选择保留，等于贪婪移除效用最低的)
B_Final = false(1, num_tiles);
AccumulatedSize = 0;

% 循环遍历排序后的瓦片，贪婪地保留它们
for idx_g = Sorted_Indices_Global
    
    % (A) 检查是否可以保留该瓦片 (贪婪保留逻辑)
    if AccumulatedSize + Tile_Size(idx_g) <= C
        % 尝试将当前瓦片加入最终集合
        B_Final(idx_g) = true;
        AccumulatedSize = AccumulatedSize + Tile_Size(idx_g);
        
        % (B) 强制保留所有依赖瓦片 (如果它们在原始 MWC 结果中)
        % MWC 结果 B_MWC 已经保证了依赖，但削减过程中可能打乱
        % 这里的削减逻辑是：按密度从高到低决定哪些**保留**，**保留**瓦片A时，必须同时保留瓦片B。
        % 但是，更符合论文描述的削减步骤是：从低密度开始**移除**。
        
        % --- 修正为更符合论文的“贪婪移除”逻辑 ---
        
        % 既然我们是按降序遍历，应该采取“贪婪保留”的逻辑，并确保依赖。
        % 由于 MWC 已经保证了依赖，现在只需确保保留的集合满足容量 C 即可。
        % 如果瓦片 A 被保留，但它依赖的瓦片 B (A -> B) 却因为密度低而被移除，就会违反约束。
        
        % 为了简化且满足约束：我们只需要确保一旦保留了 A，其所有父瓦片 B 也被保留。
        
        % 查找当前瓦片 idx_g 依赖的所有父瓦片（被依赖方）
        Parent_Tiles = find(dependency_matrix(idx_g, :) == 1);
        
        % 检查保留 idx_g 是否会引入新的依赖瓦片导致超限
        size_to_add = Tile_Size(idx_g);
        New_Parents_to_add = [];
        
        for parent_idx = Parent_Tiles
            % 检查父瓦片是否在 MWC 集合中，但尚未被保留到 B_Final 中
            if B_MWC(parent_idx) && ~B_Final(parent_idx)
                New_Parents_to_add = [New_Parents_to_add, parent_idx];
                size_to_add = size_to_add + Tile_Size(parent_idx);
            end
        end
        
        % 最终检查：保留当前瓦片和其所有必须保留的父瓦片，是否超限
        if AccumulatedSize - Tile_Size(idx_g) + size_to_add <= C 
            % 如果不超限，则正式保留当前瓦片
            B_Final(idx_g) = true;
            AccumulatedSize = AccumulatedSize - Tile_Size(idx_g) + size_to_add; % 更新大小
            
            % 强制保留所有新的父瓦片
            for parent_idx = New_Parents_to_add
                B_Final(parent_idx) = true;
            end
        else
            % 如果超限，则跳过当前瓦片，不保留
            B_Final(idx_g) = false;
        end
        
    end
end

% 确保最终集合中的瓦片都在原始 MWC 集合中，并且所有依赖都满足
B_Final = B_Final & B_MWC; 

% 重新计算最终大小
FinalSize = sum(B_Final .* Tile_Size);
disp(['削减后最终容量: ', num2str(FinalSize), ' (容量C=', num2str(C), ')']);

end

% --- 辅助函数 (需自行实现或提供) ---
function psi_out = ProbabilityDistribution(E, X)
    % 假设这是一个返回 1x(E*X) 随机概率向量的占位函数
    rng(100); % 固定随机数，便于测试
    psi_out = rand(1, E * X);
end

File: algorithm/capacity_refinement.py
"""
Capacity_Refinement - Stage 3: dependency-aware reduction when capacity is
exceeded (single-RSU version).

Input:
    B_MWC: 1-D bool array, MWC algorithm's optimal set (may exceed capacity)
    W_net: 1-D array of net expected utility
    Tile_Size: 1-D array of tile sizes (typically all 1)
    dependency_matrix: (n, n) matrix, dependency_matrix[i, j] == 1 means
                       tile i depends on tile j
    C: scalar, maximum storage capacity of the RSU

Output:
    B_Final: 1-D bool array, final cache decisions satisfying capacity
             and dependency constraints

Algorithm:
    If the current cached set exceeds capacity C, tiles are greedily retained
    in descending order of utility density Psi[b] = W_net[b] / Tile_Size[b].
    When retaining a tile, all its prerequisite parent tiles (that were in the
    original MWC set) are also retained to satisfy dependencies.
"""

import numpy as np


def capacity_refinement(B_MWC, W_net, Tile_Size, dependency_matrix, C):
    """
    Refine cache decisions to satisfy single-RSU capacity constraints.

    Parameters
    ----------
    B_MWC : np.ndarray
        1-D bool array, initial MWC cache decisions.
    W_net : np.ndarray
        1-D float array, net expected utility per tile.
    Tile_Size : np.ndarray
        1-D float array, size of each tile.
    dependency_matrix : np.ndarray
        (n, n) matrix of dependencies.
    C : float
        Maximum storage capacity.

    Returns
    -------
    B_Final : np.ndarray
        1-D bool array, refined cache decisions.
    """
    num_tiles = len(W_net)
    B_Current = B_MWC.copy()
    CurrentSize = np.sum(B_Current * Tile_Size)

    if CurrentSize <= C:
        return B_Current

    # ---- 1. Compute utility density Psi[b] = W_net[b] / Tile_Size[b] ----
    Psi = W_net / Tile_Size

    # ---- 2. Identify tiles currently cached ----
    Indices_To_Consider = np.where(B_Current)[0]
    Psi_to_sort = Psi[Indices_To_Consider]

    # ---- 3. Sort by utility density descending ----
    sorted_local = np.argsort(-Psi_to_sort)  # descending
    Sorted_Indices_Global = Indices_To_Consider[sorted_local]

    # ---- 4. Dependency-aware greedy retention ----
    B_Final = np.zeros(num_tiles, dtype=bool)
    AccumulatedSize = 0.0

    for idx_g in Sorted_Indices_Global:
        # Check if we can retain this tile
        if AccumulatedSize + Tile_Size[idx_g] <= C:
            # Tentatively retain
            B_Final[idx_g] = True
            AccumulatedSize += Tile_Size[idx_g]

            # Find all parent tiles that this tile depends on
            Parent_Tiles = np.where(dependency_matrix[idx_g, :] == 1)[0]

            # Check if retaining parents would exceed capacity
            size_to_add = Tile_Size[idx_g]
            New_Parents_to_add = []

            for parent_idx in Parent_Tiles:
                # Parent must be in original MWC set but not yet retained
                if B_MWC[parent_idx] and not B_Final[parent_idx]:
                    New_Parents_to_add.append(parent_idx)
                    size_to_add += Tile_Size[parent_idx]

            # Final check: can we fit this tile + its newly needed parents?
            if AccumulatedSize - Tile_Size[idx_g] + size_to_add <= C:
                # Yes — retain tile and its parents
                B_Final[idx_g] = True
                AccumulatedSize = (AccumulatedSize - Tile_Size[idx_g]
                                   + size_to_add)
                for parent_idx in New_Parents_to_add:
                    B_Final[parent_idx] = True
            else:
                # No — skip this tile
                B_Final[idx_g] = False
                AccumulatedSize -= Tile_Size[idx_g]

    # Ensure final set is a subset of original MWC set
    B_Final = B_Final & B_MWC

    FinalSize = np.sum(B_Final * Tile_Size)
    print(f"  削减后最终容量: {FinalSize:.0f} (容量C={C:.0f})")

    return B_Final
